File: slidingdrawer/src/main/java/com/pierfrancescosoffritti/slidingdrawer/SlidingPanel.java
package com.pierfrancescosoffritti.slidingdrawer;

import android.animation.ValueAnimator;
import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewConfiguration;
import android.view.ViewGroup;
import android.view.animation.DecelerateInterpolator;
import android.widget.LinearLayout;

import java.util.HashSet;
import java.util.Set;

/**
 * Custom View implementing a bottom sheet that is part of the view hierarchy, not above it.
 * <br/><br/>
 * This ViewGroup can have only 2 children:
 * <ol>
 * <li>The first children is the non slidable view (the view that will be covered when the bottom sheet is expanded).</li>
 * <li>The seconds children is the slidable view (the actual bottom sheet, the view that will be sliding over the non slidable view).</li>
 * </ol>
 */
public class SlidingPanel extends LinearLayout {

    private final int TOUCH_SLOP = ViewConfiguration.get(getContext()).getScaledTouchSlop();
    private static final int SLIDE_DURATION_SHORT = 300;
    private static final int SLIDE_DURATION_LONG = 600;

    // the color of the shade that fades over the non slidable view when the slidable view slides
    private static final int SHADE_COLOR_WITH_ALPHA = 0x99000000;
    private static final int SHADE_COLOR_MAX_ALPHA = SHADE_COLOR_WITH_ALPHA >>> 24;
    private static final int SHADE_COLOR = SHADE_COLOR_WITH_ALPHA & 0x00FFFFFF;

    // view that will slide
    private View slidingView;

    // view that won't slide
    private View nonSlidingView;

    // the only view sensible to vertical dragging. Dragging this view will slide the slidingView
    private View dragView;

    // an optional view, child of slidingView, to which a margin bottom is added, in order to make it always fit in the screen.
    private View fittingView;

    private PanelState state = PanelState.COLLAPSED;
    // A value between 1.0 and 0.0 (1.0 = EXPANDED, 0.0 = COLLAPSED)
    private float currentSlide = 0.0f;

    // slidingView coordinate when expanded
    private int maxSlide;
    // slidingView coordinate when collapsed
    private final int minSlide = 0;

    // duration of the slide in milliseconds, when executed with an animation instead of a gesture
    private long slideDuration = SLIDE_DURATION_SHORT;

    private float slidingViewPositioOnTouchDown;
    private float initialTouchCoordinates;

    boolean isSliding = false;

    private final Drawable elevationShadow90Deg;
    private final Drawable elevationShadow180Deg;
    private int elevationShadowLength;

    private int slidingViewId;
    private int nonSlidingViewId;
    private int dragViewId;
    private int fittingViewId;

    private boolean fitSlidingContentToScreen;

    private final Set<OnSlideListener> listeners = new HashSet<>();

    public SlidingPanel(Context context) {
        this(context, null);
    }

    public SlidingPanel(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public SlidingPanel(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        applyAttributes(attrs);

        elevationShadow90Deg = ContextCompat.getDrawable(getContext(), R.drawable.elevation_shadow_90_deg);
        elevationShadow180Deg = ContextCompat.getDrawable(getContext(), R.drawable.elevation_shadow_180_deg);
        setWillNotDraw(false);
    }

    private void applyAttributes(AttributeSet attrs) {
        TypedArray typedArray = getContext().getTheme().obtainStyledAttributes(attrs, R.styleable.SlidingPanel, 0, 0);

        try {
            slidingViewId = typedArray.getResourceId(R.styleable.SlidingPanel_slidingView, -1);
            nonSlidingViewId = typedArray.getResourceId(R.styleable.SlidingPanel_nonSlidingView, -1);
            dragViewId = typedArray.getResourceId(R.styleable.SlidingPanel_dragView, -1);
            fittingViewId = typedArray.getResourceId(R.styleable.SlidingPanel_fitViewToScreen, -1);

            elevationShadowLength = typedArray.getDimensionPixelSize(R.styleable.SlidingPanel_elevation, 10);
            fitSlidingContentToScreen = typedArray.getBoolean(R.styleable.SlidingPanel_fitSlidingContentToScreen, true);
        } finally {
            typedArray.recycle();
        }

        if(slidingViewId == -1)
            throw new RuntimeException("SlidingPanel, app:slidingView attribute not set. You must set this attribute to the id of the view that you want to be sliding.");
        if(nonSlidingViewId == -1)
            throw new RuntimeException("SlidingPanel, app:nonSlidingViewId attribute not set. You must set this attribute to the id of the view that you want to be static.");
        if(fitSlidingContentToScreen && fittingViewId != -1)
            throw new RuntimeException("SlidingPanel, app:fitSlidingContentToScreen is set to true and app:fitViewToScreen is used." +
                    " This two attributes are mutually exclusive, you can use only one at a time.");
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();

        if (getChildCount() != 2)
            throw new IllegalStateException("SlidingPanel must have exactly 2 children. But has " +getChildCount());

        slidingView = findViewById(slidingViewId);
        nonSlidingView = findViewById(nonSlidingViewId);
        dragView = findViewById(dragViewId);

        fittingView = findViewById(fittingViewId);

        if(slidingView == null)
            throw new RuntimeException("SlidingPanel, slidingView is null.");
        if(nonSlidingView == null)
            throw new RuntimeException("SlidingPanel, nonSlidingView is null.");

        // optional views
        if(dragViewId != -1 && dragView == null)
            throw new RuntimeException("SlidingPanel, can't find dragView.");
        if(fittingViewId != -1 && fittingView == null)
            throw new RuntimeException("SlidingPanel, can't find fittingView.");

        if(dragView == null)
            dragView = slidingView;
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent currentTouchEvent) {
        if(!Utils.INSTANCE.isMotionEventWithinBoundaries(currentTouchEvent, dragView)) {
            isSliding = false;
            return false;
        }

        switch (currentTouchEvent.getAction()) {
            case MotionEvent.ACTION_DOWN: {
                initialTouchCoordinates = getOrientation() == VERTICAL ? currentTouchEvent.getY() : currentTouchEvent.getX();
                slidingViewPositioOnTouchDown = getOrientation() == VERTICAL ? slidingView.getY() : slidingView.getX();
                return false;
            } case MotionEvent.ACTION_MOVE: {
                float currentTouch = getOrientation() == VERTICAL ? currentTouchEvent.getY() : currentTouchEvent.getX();
                final float diff = Math.abs(currentTouch - initialTouchCoordinates);

                isSliding = diff > TOUCH_SLOP/4;
                return isSliding;
            } default:
                isSliding = false;
                return false;
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    @Override
    public boolean onTouchEvent(MotionEvent touchEvent) {
        float currentTouchEvent = getOrientation() == VERTICAL ? touchEvent.getY() : touchEvent.getX();

        switch (touchEvent.getAction()) {
            case MotionEvent.ACTION_UP:
                if(state == PanelState.SLIDING)
                    completeSlide(currentSlide, currentTouchEvent > initialTouchCoordinates ? SlidingDirection.DOWN : SlidingDirection.UP);
                break;
            case MotionEvent.ACTION_MOVE:
                if(!isSliding && !Utils.INSTANCE.isMotionEventWithinBoundaries(touchEvent, dragView))
                    return false;

                float touchOffset = initialTouchCoordinates - currentTouchEvent;
                float finalPosition = slidingViewPositioOnTouchDown - touchOffset;

                if(finalPosition > maxSlide)
                    finalPosition = maxSlide;
                else if(finalPosition < minSlide)
                    finalPosition = minSlide;

                updateState(Utils.INSTANCE.normalize(finalPosition, maxSlide));
                break;
        }
        return true;
    }

    @Override
    public boolean performClick() {
        super.performClick();

        if(state == PanelState.EXPANDED) setState(PanelState.COLLAPSED); else setState(PanelState.EXPANDED);

        return true;
    }

    /**
     * Ask all children to measure themselves, then compute the measure of this layout based on its children.
     */
    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        int maxHeight = 0;
        int maxWidth = 0;
        int childState = 0;

        // Iterate through all children, measuring them while using their size to compute the dimensions of this view.
        for (int i = 0; i < getChildCount(); i++) {
            final View child = getChildAt(i);
            if (child.getVisibility() == GONE)
                continue;

            // Measure the child.
            measureChildWithMargins(child, widthMeasureSpec, 0, heightMeasureSpec, 0);

            // Update SlidingPanel size based on layout params of child.
            // Children that asked to be positioned on the left or right go in those gutters.
            final LayoutParams lp = (LayoutParams) child.getLayoutParams();
            maxWidth = Math.max(maxWidth, child.getMeasuredWidth() + lp.leftMargin + lp.rightMargin);

            maxHeight = Math.max(maxHeight, child.getMeasuredHeight() + lp.topMargin + lp.bottomMargin);
            childState = combineMeasuredStates(childState, child.getMeasuredState());
        }

        // Check against our minimum height and width
        maxHeight = Math.max(maxHeight, getSuggestedMinimumHeight());
        maxWidth = Math.max(maxWidth, getSuggestedMinimumWidth());

        // Report our final dimensions.
        setMeasuredDimension(
                resolveSizeAndState(maxWidth, widthMeasureSpec, childState),
                resolveSizeAndState(
                        maxHeight,
                        heightMeasureSpec,
                        childState << MEASURED_HEIGHT_STATE_SHIFT
                )
        );
    }

    private final Rect tmpContainerRect = new Rect();
    private final Rect tmpChildRect = new Rect();

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        int currentTop = getPaddingTop();
        int currentLeft = getPaddingLeft();

        for (int i=0; i<getChildCount(); i++) {
            final View child = getChildAt(i);

            if (child.getVisibility() == GONE)
                continue;

            final LayoutParams childLayoutParams = (LayoutParams) child.getLayoutParams();

            final int childWidth = child.getMeasuredWidth();
            final int childHeight = child.getMeasuredHeight();

            tmpContainerRect.left = currentLeft + childLayoutParams.leftMargin;
            tmpContainerRect.right = currentLeft + childWidth - childLayoutParams.rightMargin;

            tmpContainerRect.top = currentTop + childLayoutParams.topMargin;
            tmpContainerRect.bottom = currentTop + childHeight - childLayoutParams.bottomMargin;

            if(getOrientation() == VERTICAL)
                currentTop = tmpContainerRect.bottom;
            else
                currentLeft = tmpContainerRect.right;

            Gravity.apply(childLayoutParams.gravity, childWidth, childHeight, tmpContainerRect, tmpChildRect);

            child.layout(tmpChildRect.left, tmpChildRect.top, tmpChildRect.right, tmpChildRect.bottom);
        }
    }

    @Override
    public void draw(Canvas canvas) {
        super.draw(canvas);
        if(getOrientation() == VERTICAL) drawElevationShadow90(canvas); else drawElevationShadow180(canvas);
    }

    private void drawElevationShadow90(Canvas canvas) {
        final int top = (int) (slidingView.getY() - elevationShadowLength);
        final int bottom = (int) slidingView.getY();
        final int left = slidingView.getLeft();
        final int right = slidingView.getRight();

        elevationShadow90Deg.setBounds(left, top, right, bottom);
        elevationShadow90Deg.draw(canvas);
    }

    private void drawElevationShadow180(Canvas canvas) {
        final int top = slidingView.getTop();
        final int bottom = slidingView.getBottom();
        final int left = (int) (slidingView.getX() - elevationShadowLength);
        final int right = (int) slidingView.getX();

        elevationShadow180Deg.setBounds(left, top, right, bottom);
        elevationShadow180Deg.draw(canvas);
    }

    private final Rect tmpRect = new Rect();
    private final Paint shadePaint = new Paint();

    @Override
    protected boolean drawChild(Canvas canvas, View child, long drawingTime) {
        boolean result;
        final int save = canvas.save();

        if(child == nonSlidingView) {
            maxSlide = getOrientation() == VERTICAL ? nonSlidingView.getHeight() : nonSlidingView.getWidth();

            // Clip against the slider; no sense drawing what will immediately be covered,
            // Unless the panel is set to overlay content
            canvas.getClipBounds(tmpRect);
            tmpRect.bottom = Math.min(tmpRect.bottom, slidingView.getTop());

            result = super.drawChild(canvas, child, drawingTime);

            if (currentSlide > 0) {
                final int currentShadeAlpha = (int) (SHADE_COLOR_MAX_ALPHA * currentSlide);
                final int currentShadeColor = currentShadeAlpha << 24 | SHADE_COLOR;
                shadePaint.setColor(currentShadeColor);
                canvas.drawRect(tmpRect, shadePaint);
            }
        } else if (child == slidingView) {
            applyFitToScreen();
            result = super.drawChild(canvas, child, drawingTime);
        } else {
            result = super.drawChild(canvas, child, drawingTime);
        }

        canvas.restoreToCount(save);

        return result;
    }

    boolean fitScreenApplied = false;
    private void applyFitToScreen() {
        if(fitScreenApplied) return;

        int side = getOrientation() == VERTICAL ? 3 : 2;
        if (fitSlidingContentToScreen) {
            if (slidingView instanceof ViewGroup) {
                for (int i = 0; i < ((ViewGroup) slidingView).getChildCount(); i++)
                    Utils.INSTANCE.setMargin(((ViewGroup) slidingView).getChildAt(i), maxSlide, side);
            } else {
                Utils.INSTANCE.setPadding(slidingView, maxSlide, side);
            }
        } else if(fittingView != null) {
            Utils.INSTANCE.setMargin(fittingView, maxSlide, side);
        }

        fitScreenApplied = true;
    }

    /**
     * Use this method to automatically slide at a specific position
     * @param positionNormalized Normalized value (between 0.0 and 1.0) representing the final slide position
     */
    public void slideTo(float positionNormalized) {
        if(Float.isNaN(positionNormalized))
            throw new IllegalArgumentException("Bad value. Can't slide to NaN");
        if(positionNormalized < 0 || positionNormalized > 1)
            throw new IllegalArgumentException("Bad value. Can't slide to " +positionNormalized +". Value must be between 0 and 1");

        ValueAnimator va = ValueAnimator.ofFloat(currentSlide, positionNormalized);
        va.setInterpolator(new DecelerateInterpolator(1.5f));
        va.setDuration(slideDuration);
        va.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {

            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                updateState((Float) animation.getAnimatedValue());
            }
        });
        va.start();
    }

    private void completeSlide(float currentSlide, SlidingDirection direction) {
        float targetSlide;

        switch (direction) {
            case UP:
                if(currentSlide > 0.1)
                    targetSlide = minSlide;
                else
                    targetSlide = maxSlide;
                break;
            case DOWN:
                if(currentSlide < 0.9)
                    targetSlide = maxSlide;
                else
                    targetSlide = minSlide;
                break;
            default:
                return;
        }

        slideTo(Utils.INSTANCE.normalize(targetSlide, maxSlide));
    }

    /**
     * always use this method to update the position of the sliding view.
     * @param newSlideRatio new slide value, normalized between 0 and 1
     */
    private void updateState(float newSlideRatio) {
        if(newSlideRatio < 0 || newSlideRatio > 1)
            throw new IllegalArgumentException("Slide value \"" +newSlideRatio +"\" should be normalized, between 0 and 1.");

        currentSlide = newSlideRatio;
        state = currentSlide == 1 ? PanelState.EXPANDED : currentSlide == 0 ? PanelState.COLLAPSED : PanelState.SLIDING;

        float currentSlideNonNormalized = Math.abs((currentSlide * maxSlide) - maxSlide);
        if(getOrientation() == VERTICAL)
            slidingView.setY(currentSlideNonNormalized);
        else
            slidingView.setX(currentSlideNonNormalized);

        invalidate();
        notifyListeners(currentSlide);
    }

    public void setDragView(@NonNull View dragView) {
        this.dragView = dragView;
    }

    /**
     * @return shadow height in pixels
     */
    public int getElevationShadowLength() {
        return elevationShadowLength;
    }

    /**
     * @param elevationShadowLength shadow height in pixels
     */
    public void setElevationShadowLength(int elevationShadowLength) {
        this.elevationShadowLength = elevationShadowLength;
    }

    public PanelState getState() {
        return state;
    }

    /**
     * Use this method to change the state of the slidingView.
     * @param state a value from {@link PanelState}
     */
    public void setState(PanelState state) {
        if(state == this.state)
            return;

        switch (state) {
            case EXPANDED:
                slideTo(1);
                break;
            case COLLAPSED:
                slideTo(0);
                break;
            case SLIDING:
                break;
        }
    }

    public long getSlideDuration() {
        return slideDuration;
    }

    /**
     * Set the duration of the automatic slide animation
     * @param slideDuration duration in milliseconds
     */
    public void setSlideDuration(long slideDuration) {
        this.slideDuration = slideDuration;
    }

    public void addSlideListener(@NonNull OnSlideListener listener) {
        listeners.add(listener);
    }

    public void removeListener(@NonNull OnSlideListener listener) {
        listeners.remove(listener);
    }

    private void notifyListeners(float currentSlide) {
        for(OnSlideListener listener : listeners)
            listener.onSlide(this, currentSlide);
    }

    /**
     * Implement this interface if you want to observe slide changes
     */
    public interface OnSlideListener {
        void onSlide(SlidingPanel slidingPanel, float currentSlide);
    }
}

File: slidingdrawer/src/main/java/com/pierfrancescosoffritti/slidingdrawer/SlidingPanel.kt
package com.pierfrancescosoffritti.slidingdrawer

import android.animation.ValueAnimator
import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import androidx.core.content.ContextCompat
import android.util.AttributeSet
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.view.ViewConfiguration
import android.view.ViewGroup
import android.view.animation.DecelerateInterpolator
import android.widget.LinearLayout

import com.pierfrancescosoffritti.slidingdrawer.utils.Utils
import com.pierfrancescosoffritti.slidingdrawer.utils.ViewUtils
import com.pierfrancescosoffritti.slidingdrawer.utils.Extensions.isOrientationVertical
import com.pierfrancescosoffritti.slidingdrawer.utils.Extensions.isMotionEventWithinBounds
import java.lang.Float.isNaN

import java.util.HashSet

/**
 * Custom View implementing a sliding panel (bottom sheet pattern) that is part of the view hierarchy, not above it.
 *
 * Read [detailed documentation here](https://github.com/PierfrancescoSoffritti/sliding-drawer)
 */
class SlidingPanel(context: Context, attrs: AttributeSet? = null) : LinearLayout(context, attrs) {

    companion object {
        const val  SLIDE_DURATION_SHORT = 300L
        const val  SLIDE_DURATION_LONG = 600L

        // the color of the shade that fades over the nonSlidingView view when the slidingView slides over it
        private const val SHADE_COLOR_WITH_ALPHA = -0x67000000
        private const val SHADE_COLOR_MAX_ALPHA = SHADE_COLOR_WITH_ALPHA.ushr(24)
        private const val SHADE_COLOR = SHADE_COLOR_WITH_ALPHA and 0x00FFFFFF
    }

    private val touchSlop = ViewConfiguration.get(getContext()).scaledTouchSlop

    private lateinit var slidingView: View
    private lateinit var nonSlidingView: View

    // the view sensible to swipe gestures. Drag this view to move the slidingView
    private lateinit var dragView: View

    // an optional view, child of slidingView, to which a margin bottom is added, in order to make it always fit on the screen.
    private var fittingView: View? = null

    private var state = PanelState.COLLAPSED
    // A value between 1.0 and 0.0 (0.0 = COLLAPSED, 1.0 = EXPANDED)
    private var currentSlide = 0.0f

    // the maximum amount the slidingView can slide. It corresponds to the height (width) of the nonSlidingView.
    private var maxSlide = 0f
    // the minimum coordinate the slidingView can slide to. It corresponds to the top (right) of the nonSlidingView.
    private var minSlide = 0f

    // position of the slidingView recorded when the user touches the screen for the first time (eg. at the beginning of a swipe gesture)
    private var slidingViewPosAtFirstTouch = 0f
    // touch coordinates of the first touch gesture (eg. at the beginning of a swipe gesture)
    private var coordOfFirstTouch = 0f

    private var isSliding = false

    private val elevationShadow90Deg = ContextCompat.getDrawable(getContext(), R.drawable.elevation_shadow_90_deg)!!
    private val elevationShadow180Deg = ContextCompat.getDrawable(getContext(), R.drawable.elevation_shadow_180_deg)!!

    private val slidingViewId: Int
    private val nonSlidingViewId: Int
    private val fittingViewId: Int
    private var dragViewId: Int

    private val fitSlidingViewContentToScreen: Boolean
    private var fitScreenApplied = false

    private val listeners = HashSet<OnSlideListener>()

    /**
     * The duration of the slide in millisecond, when executed with an animation instead of a gesture
     */
    var slideDuration = SLIDE_DURATION_SHORT

    /**
     * Sliding panel shadow height in pixels
     */
    var elevationShadowLength: Int = 0

    init {
        setWillNotDraw(false)

        val typedArray = context.theme.obtainStyledAttributes(attrs, R.styleable.SlidingPanel, 0, 0)

        try {
            slidingViewId = typedArray.getResourceId(R.styleable.SlidingPanel_slidingView, -1)
            nonSlidingViewId = typedArray.getResourceId(R.styleable.SlidingPanel_nonSlidingView, -1)
            dragViewId = typedArray.getResourceId(R.styleable.SlidingPanel_dragView, -1)
            fittingViewId = typedArray.getResourceId(R.styleable.SlidingPanel_fitViewToScreen, -1)

            elevationShadowLength = typedArray.getDimensionPixelSize(R.styleable.SlidingPanel_elevation, 10)
            fitSlidingViewContentToScreen = typedArray.getBoolean(R.styleable.SlidingPanel_fitSlidingContentToScreen, true)
        } finally {
            typedArray.recycle()
        }

        if (slidingViewId == -1)
            throw RuntimeException("SlidingPanel, app:slidingView attribute not set. You must set this attribute to the id of the view that you want to be sliding.")
        if (nonSlidingViewId == -1)
            throw RuntimeException("SlidingPanel, app:nonSlidingViewId attribute not set. You must set this attribute to the id of the view that you want to be static.")
        if (fitSlidingViewContentToScreen && fittingViewId != -1)
            throw RuntimeException(
                    "SlidingPanel, app:fitSlidingViewContentToScreen is set to true and app:fitViewToScreen is used."
                            +" This two attributes are mutually exclusive, you can use only one at a time."
            )
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()

        if (childCount != 2)
            throw IllegalStateException("SlidingPanel must have exactly 2 children. But has $childCount")

        slidingView = findViewById(slidingViewId) ?: throw RuntimeException("SlidingPanel, slidingView is null.")
        nonSlidingView = findViewById(nonSlidingViewId) ?: throw RuntimeException("SlidingPanel, nonSlidingView is null.")

        dragView = findViewById(dragViewId) ?:
                if (dragViewId != -1) throw RuntimeException("SlidingPanel, can't find dragView.")
                else slidingView

        fittingView = findViewById(fittingViewId) ?:
                if(fittingViewId != -1) throw RuntimeException("SlidingPanel, can't find fittingView.")
                else null
    }

    override fun onInterceptTouchEvent(currentTouchEvent: MotionEvent): Boolean {
        if (!dragView.isMotionEventWithinBounds(currentTouchEvent)) {
            isSliding = false
            return false
        }

        when (currentTouchEvent.action) {
            MotionEvent.ACTION_DOWN -> {
                coordOfFirstTouch = if (isOrientationVertical()) currentTouchEvent.y else currentTouchEvent.x
                slidingViewPosAtFirstTouch = if (isOrientationVertical()) slidingView.y else slidingView.x
                return false
            }
            MotionEvent.ACTION_MOVE -> {
                val currentTouch: Float = if (isOrientationVertical()) currentTouchEvent.y else currentTouchEvent.x
                val diff: Float = Math.abs(currentTouch - coordOfFirstTouch)
                isSliding = diff > touchSlop / 4
                return isSliding
            }
            else -> {
                isSliding = false
                return false
            }
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(touchEvent: MotionEvent): Boolean {
        val currentTouchEvent = if (isOrientationVertical()) touchEvent.y else touchEvent.x

        when (touchEvent.action) {
            MotionEvent.ACTION_UP ->
                if (state == PanelState.SLIDING)
                    completeSlide(currentSlide, if (currentTouchEvent > coordOfFirstTouch) SlidingDirection.DOWN else SlidingDirection.UP)
            MotionEvent.ACTION_MOVE -> {
                if (!isSliding && !dragView.isMotionEventWithinBounds(touchEvent))
                    return false

                val touchOffset: Float = coordOfFirstTouch - currentTouchEvent
                var finalPosition: Float = slidingViewPosAtFirstTouch - touchOffset

                finalPosition = Utils.clamp(finalPosition, minSlide, maxSlide)
                updateState(Utils.normalizeScreenCoordinate(finalPosition, maxSlide))
            }
        }
        return true
    }

    override fun performClick(): Boolean {
        super.performClick()
        toggleState()
        return true
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        var slidingPanelHeight = 0
        var slidingPanelWidth = 0
        var childrenCombinedMeasuredStates = 0

        for (i: Int in 0 until childCount) {
            val child: View = getChildAt(i)
            if (child.visibility == View.GONE)
                continue

            measureChildWithMargins(child, widthMeasureSpec, 0, heightMeasureSpec, 0)

            val layoutParams = child.layoutParams as LinearLayout.LayoutParams

            slidingPanelWidth = Math.max(slidingPanelWidth, child.measuredWidth + layoutParams.leftMargin + layoutParams.rightMargin)
            slidingPanelHeight = Math.max(slidingPanelHeight, child.measuredHeight + layoutParams.topMargin + layoutParams.bottomMargin)

            childrenCombinedMeasuredStates = View.combineMeasuredStates(childrenCombinedMeasuredStates, child.measuredState)
        }

        slidingPanelHeight = Math.max(slidingPanelHeight, suggestedMinimumHeight)
        slidingPanelWidth = Math.max(slidingPanelWidth, suggestedMinimumWidth)

        setMeasuredDimension(
                View.resolveSizeAndState(slidingPanelWidth, widthMeasureSpec, childrenCombinedMeasuredStates),
                View.resolveSizeAndState(slidingPanelHeight, heightMeasureSpec, childrenCombinedMeasuredStates)
        )
    }

    private val onLayoutContainerRect = Rect()
    private val onLayoutChildRect = Rect()

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        var currentTop = paddingTop
        var currentLeft = paddingLeft

        for (i: Int in 0 until childCount) {
            val child: View = getChildAt(i)

            if (child.visibility == View.GONE)
                continue

            val childLayoutParams = child.layoutParams as LinearLayout.LayoutParams

            val childWidth = child.measuredWidth
            val childHeight = child.measuredHeight

            onLayoutContainerRect.left = currentLeft + childLayoutParams.leftMargin
            onLayoutContainerRect.right = currentLeft + childWidth - childLayoutParams.rightMargin

            onLayoutContainerRect.top = currentTop + childLayoutParams.topMargin
            onLayoutContainerRect.bottom = currentTop + childHeight - childLayoutParams.bottomMargin

            if (isOrientationVertical())
                currentTop = onLayoutContainerRect.bottom
            else
                currentLeft = onLayoutContainerRect.right

            Gravity.apply(childLayoutParams.gravity, childWidth, childHeight, onLayoutContainerRect, onLayoutChildRect)

            child.layout(onLayoutChildRect.left, onLayoutChildRect.top, onLayoutChildRect.right, onLayoutChildRect.bottom)
        }
    }

    override fun draw(canvas: Canvas) {
        super.draw(canvas)
        if (isOrientationVertical())
            drawElevationShadow90(canvas)
        else
            drawElevationShadow180(canvas)
    }

    private fun drawElevationShadow90(canvas: Canvas) {
        val top = (slidingView.y - elevationShadowLength).toInt()
        val bottom = slidingView.y.toInt()
        val left = slidingView.left
        val right = slidingView.right

        elevationShadow90Deg.setBounds(left, top, right, bottom)
        elevationShadow90Deg.draw(canvas)
    }

    private fun drawElevationShadow180(canvas: Canvas) {
        val top = slidingView.top
        val bottom = slidingView.bottom
        val left = (slidingView.x - elevationShadowLength).toInt()
        val right = slidingView.x.toInt()

        elevationShadow180Deg.setBounds(left, top, right, bottom)
        elevationShadow180Deg.draw(canvas)
    }

    private val drawChildChildTempRect = Rect()
    private val shadePaint = Paint()

    override fun drawChild(canvas: Canvas, child: View, drawingTime: Long): Boolean {
        val save = canvas.save()
        val result: Boolean

        if (child === nonSlidingView) {
            maxSlide = if (isOrientationVertical()) nonSlidingView.height.toFloat() else nonSlidingView.width.toFloat()
            minSlide = if (isOrientationVertical()) nonSlidingView.top.toFloat() else nonSlidingView.right.toFloat()

            canvas.getClipBounds(drawChildChildTempRect)
            result = super.drawChild(canvas, child, drawingTime)

            if (currentSlide > 0) {
                val currentShadeAlpha = (SHADE_COLOR_MAX_ALPHA * currentSlide).toInt()
                val currentShadeColor = currentShadeAlpha shl 24 or SHADE_COLOR
                shadePaint.color = currentShadeColor
                canvas.drawRect(drawChildChildTempRect, shadePaint)
            }
        } else if (child == slidingView) {
            applyFitToScreenOnce()
            result = super.drawChild(canvas, child, drawingTime)
        } else {
            result = super.drawChild(canvas, child, drawingTime)
        }

        canvas.restoreToCount(save)
        return result
    }

    private fun applyFitToScreenOnce() {
        if (fitScreenApplied) return

        val side = if (isOrientationVertical()) Side.BOTTOM else Side.RIGHT
        if (fitSlidingViewContentToScreen) {
            if (slidingView is ViewGroup) {
                for (i: Int in 0 until (slidingView as ViewGroup).childCount)
                    ViewUtils.setMargin((slidingView as ViewGroup).getChildAt(i), maxSlide.toInt(), side)
            } else {
                ViewUtils.setPadding(slidingView, maxSlide.toInt(), side)
            }
        } else if (fittingView != null) {
            ViewUtils.setMargin(fittingView!!, maxSlide.toInt(), side)
        }

        fitScreenApplied = true
    }

    /**
     * Use this method to automatically slide at a specific position
     * @param positionNormalized Normalized value (between 0.0 and 1.0) representing the final slide position
     */
    private fun slideTo(positionNormalized: Float) {
        if (isNaN(positionNormalized))
            throw IllegalArgumentException("Bad value. Can't slide to NaN")
        if (positionNormalized < 0 || positionNormalized > 1)
            throw IllegalArgumentException("Bad value. Can't slide to $positionNormalized. Value must be between 0 and 1")

        val va = ValueAnimator.ofFloat(currentSlide, positionNormalized)
        va.interpolator = DecelerateInterpolator(1.5f)
        va.duration = slideDuration
        va.addUpdateListener { animation -> updateState(animation.animatedValue as Float) }
        va.start()
    }

    private fun completeSlide(currentSlide: Float, direction: SlidingDirection) {
        val targetSlide: Float = when (direction) {
            SlidingDirection.UP -> if (currentSlide > 0.1)
                minSlide
            else
                maxSlide
            SlidingDirection.DOWN -> if (currentSlide < 0.9)
                maxSlide
            else
                minSlide
            else -> return
        }

        slideTo(Utils.normalizeScreenCoordinate(targetSlide, maxSlide))
    }

    /**
     * always use this method to update the position of the sliding view.
     * @param newSlideRatio new slide value, normalized between 0 and 1
     */
    private fun updateState(newSlideRatio: Float) {
        if (newSlideRatio < 0 || newSlideRatio > 1)
            throw IllegalArgumentException("Slide value \"$newSlideRatio\" should be normalized, between 0 and 1.")

        currentSlide = newSlideRatio
        state = if (currentSlide == 1f) PanelState.EXPANDED else if (currentSlide == 0f) PanelState.COLLAPSED else PanelState.SLIDING

        val currentSlideNonNormalized = Math.abs(currentSlide * maxSlide - maxSlide)
        if (isOrientationVertical())
            slidingView.y = currentSlideNonNormalized
        else
            slidingView.x = currentSlideNonNormalized

        invalidate()
        notifyListeners(currentSlide)
    }

    fun setDragView(dragView: View) {
        this.dragView = dragView
        dragViewId = dragView.id
    }

    fun getState(): PanelState {
        return state
    }

    /**
     * Use this method to change the state of the slidingView.
     */
    fun setState(state: PanelState) {
        if (state === this.state)
            return

        when (state) {
            PanelState.EXPANDED -> slideTo(1f)
            PanelState.COLLAPSED -> slideTo(0f)
            PanelState.SLIDING -> {
            }
        }
    }

    fun toggleState() {
        if (state === PanelState.EXPANDED) setState(PanelState.COLLAPSED) else setState(PanelState.EXPANDED)
    }

    fun addSlideListener(listener: OnSlideListener) {
        listeners.add(listener)
    }

    fun addSlideListener(callback: (slidingPanel: SlidingPanel, state: PanelState, currentSlide: Float) -> Unit) {
        addSlideListener(object : OnSlideListener {
            override fun onSlide(slidingPanel: SlidingPanel, state: PanelState, currentSlide: Float) = callback(slidingPanel, state, currentSlide)
        })
    }

    fun removeListener(listener: OnSlideListener) {
        listeners.remove(listener)
    }

    private fun notifyListeners(currentSlide: Float) {
        for (listener in listeners)
            listener.onSlide(this, state, currentSlide)
    }

    /**
     * Implement this interface if you want to observe changes in the panel.
     */
    interface OnSlideListener {
        fun onSlide(slidingPanel: SlidingPanel, state: PanelState, currentSlide: Float)
    }
}